### DISCUSSION TITLE
Is it possible to exclude `/public/system/cache` from backup?

### DISCUSSION THREAD

**Discussion by ccoenen** [mastodon/mastodon#21287] [Q&A]
I have not seen this mentioned in the [backup docs](https://docs.joinmastodon.org/admin/backups/), so I would like to know your opinion about it. By far the largest amount of space (at least on my instance) is taken up by `/public/system/cache`. To me, the name "cache" suggests that this can be generated from something or could be obtained from other servers if need be.

**Is it safe to exclude `public/system/cache` from backup?** Would excluding it lose anyone's personal data that would be unique to my instance? Are cached files somehow also tracked in the database (which then might be out of sync with the actual directory on disk)?

**Comment by ISibboI** [comment]
I recently deleted my `public/system/cache` recently, and since then many profile pictures of foreign instance users are missing...

**Reply by bobhy** [reply]
If you do exclude the cache, you'll at least have a functioning system and your local users won't lose their data when you restore from the backup.  For a small instance, excluding the cache can shrink the amount of data to back up nearly 90% (on my small instance, from 30GB to 3), so it's an appealing idea.

I believe normal user action can restore media attachments (as users view the toots), but not the profile pics?  From the admin  command line you can do both: `tootctl media refresh` and `tootctl accounts refresh <specificBrokenUser>` (see #20881).

Also see #21066 which implements new `tootctl media {prune,refresh}-profile-media` commands which make recovering *all* the user profiles possible in one operation.  I manually hacked my instance with this PR and it seemed to work, FWIW.  If you like the idea, agitate for devs to merge it -- it seems to be good to go.

**Comment by PhilipMay** [comment]
This is a very good question. I am also thinking to exclude cache from backup.
Do you have any new experience with this?

**Reply by ccoenen** [reply]
Not really, to be honest.

**Comment by themullinator** [comment]
Yes, cached files are also tracked in the database (if you inspect things like profile images, custom emojis, and media attachments, you'll note they're all served from your domain even though the originals may be on someone else's server. That correlation has to be maintained somewhere.)

Losing your cache isn't the _worst_ thing in the world, but can be resource-intensive to recover if you care about user experience. Running `tootctl accounts refresh --all` and `tootctl media refresh --force` should fix most broken image problems. Over time, the experience should start to self-heal as well when new profile images and posts are fetched from remote servers as well.

**Comment by nightpool** [comment]
- Is it safe to exclude public/system/cache from backup?
    - No, broadly, public/system/cache represents all of the media sent to you by other servers. Unless you want to throw that data away, you should include it in backups. That said, the data never changes, so I would only recommend backing it up once, and then only backing up new files after that. Generally, I would not recommend keeping this data on a disk at all for almost any non-trivially sized instance. Instead, I would recommending using a cheap object storage solution like R2, Backblaze, or DigitalOcean buckets. This data can also be fetched dynamically from the remote server, but I wouldn't necessarily recommend that, because remote servers go down and lose data.
 - Would excluding it lose anyone's personal data that would be unique to my instance?
   - No, all data in public/system/cache is remote data.
 - Are cached files somehow also tracked in the database (which then might be out of sync with the actual directory on disk)?
   - Yes, this data is tracked in the media_attachments table. If you delete the files from the disk without marking the row as "freed", then it won't know to fetch the data from the remote server dynamically when the file is requested.

**Comment by bobhy** [comment]
As others have pointed out, the cache is data that is recoverable from somewhere else if the local copy goes missing.  It does not include anything irreplaceable.  If your server happens to lose the cached copy *and* one of your users happens to refer to it later, your server will fetch another copy from the source server.  This costs a little bit of overhead to each server and to the network path in between.  But what are the odds you're going to incur that extra cost?  

It depends on the odds of a server crash causing data loss in the first place (which is really quite rare), and on the odds of one of your users happening to refer to the data after you restored the core server (which odds decrease rapidly the older the data is) and, this is the kicker, will not be zero even if you are maintaining perfect backups of the cache anyway.

Mastodon is constantly deleting old cached data automatically! I believe the default settings are 14 days for media and longer for content and even longer for remote user profiles.  It's doing this to make room for the constant flood of newer information from other servers, so even with perfect backups of the cache, it is by no means an unusual thing for your server to be asking for a second copy of data that it had cached once.

I'm running a small (< 10 user) instance and have made the decision to not back up cached data.  This was an easy call for me.

**Reply by bobhy** [reply]
I need to clarify... The above applies to toot media and content and to user profile *content* as written.  But I do not believe it applies to profile *media* (like avatar images).  So if you lose the cached info on all the remote users that your instance knows and you do restore the profile content (or let user references restore them as needed), those profiles will look broken.  
This is arguably a bug in design (not implementation) of Mastodon (it knows the server the remote user is located on and can automatically restore the profile content from it.  Why can't it also restore profile media? Maybe a flaw in the underlying specifications?) In any case, giving the admin an option to restore profile media (and heal the broken remote profiles) is the point of #21066 that I mentioned above.  
It didn't change my calculus.  I don't back up any of the cache.  If I lose the server and if one of my few users asks me to, I'll run the extended command to restore the profile media (if they ask me real nice...)

**Comment by soletan** [comment]
Hm, I get the point about losing remote data and how this is affecting user experience. However, in my case the **public/system/cache** has about 5.2 million files/folders in it. Backing it up is a total no-go for any tool trying to inspect that folder literally takes several minutes to complete. Currently, I'm trying to switch to S3, but uploading stuff is far from processing a significant amount of those files within 12h. I doubt this is going to work without shutting down the instance for a week or two so that no new data is popping up while trying to move everything.

Maybe this is related to an issue we've encountered month ago but failed to solve: running `tootctl media remove` and `tootctl preview_cards remove` doesn't work from cron for no obvious reason, I don't know if it ever did. Usually, output of cronjobs is mailed to us. There hasn't ever been any such mail. Syslog is showing up their invocation right on time as scheduled. Running them manually from time to time is removing a significant number of files, though I don't know if it's covering **public/system/cache** or just those other folders next to it which have reasonable sizes. 

So, AFAIU, there is no way to safely clean or at least crop the cache a bit, is it? Or is it just our instance which has gone mad? I mean, I've queried media_attachments table via SQL and it has about 1.5 million entries in total. Is it reasonable to assume, that there are about 1.5 millions files in **public/system/cache** spread over 3.7 million folders? Focusing on those with created_at being older than 7 days, there are 

* 14k entries with empty remote_url, every one with a file_file_name set, 
* 1.4m entries with non-empty remote_url out of which 1.36m entries don't have file_file_name set.

How is **public/system/cache** related to anything like those entries without file_file_name being set? Oldest entry of those has been created on 2022-04-26. Latest one has been created a few seconds ago. I doubt Mastodon is properly tracking relationships of cached files here unless I completely miss the way the cache is managed.

Used queries have been:

* `select count(*) from media_attachments;`
* `select count(*) from media_attachments where remote_url='' and created_at<now() - interval '7 days';`
* `select count(*) from media_attachments where remote_url<>'' and created_at<now() - interval '7 days';`
* `select count(*) from media_attachments where remote_url<>'' and created_at<now() - interval '7 days' and file_file_name is null;`

Just to mention that: the public/system folder has total size of 280GB. Running those tools is trimming it down to about 260GB. media_attachments consumes 11GB, accounts has 333MB and site_uploads has 3.9MB, so cache takes about 245GB after trimming. Doesn't sound sane to me.

**Reply by ClearlyClaire** [reply]
> Running them manually from time to time is removing a significant number of files, though I don't know if it's covering **public/system/cache** or just those other folders next to it which have reasonable sizes.

`tootctl media remove` should affect `public/system/cache` almost exclusively (only things it should affect otherwise are remote media received before running Mastodon 3.1.4).

It doesn't cover *all* of those files, though, it only covers `public/system/cache/media_attachments` (with the upcoming release adding a `--prune-profiles` switch to also cover `public/system/cache/accounts`, and the current `tootctl preview-cards remove` taking care of `public/system/cache/preview_cards`).

It does indeed seems unlikely that most of the 245GB would be taken up by avatars, account headers and link previews, but it is worth checking nonetheless.

It is also possible there's been a bug in Mastodon, one of its dependencies, or your setup, that would for some reason have failed to remove files that are not registered anymore. In this case, `tootctl media remove-orphans` should help.

In any case, the current tooling around deleting cached data and recovering from a cleared cache is unfortunately lacking. It's slowly improving but there's still a lot to do.

**Reply by soletan** [reply]
Ah, haven't ever tried to look into the public/system/cache folder to realize there are similar subfolders as in public/system. This way, the code of remove() in media_cli.rb makes more sense with regard to using ".cached" dataset. I'll give remove-orphans another try right away. Thanks for that hint.

**Reply by soletan** [reply]
Running tootctl media remove-orphans took 45min to remove 1767 files consuming 53MB. That's pretty far from 245GB of total consumption. But I've split stats of folder sizes a bit:

| folder | size | files | last change >7days ago | remarks |
|---|--:|--:|--:|---|
| public/system/cache/accounts | 92GB | 592.909 | 571.018 | that's mostly on remote accounts for our instance has about 2500 accounts, only |
| public/system/cache/custom_emojis | 1.5GB | 86.888 | 83.142 |  |
| public/system/cache/media_attachments | 162GB | 344.969 | 55.278 | ID prefixes 108 and 109, only |
| public/system/cache/preview_cards | 5.4GB | 166.228 | 72.164 |  |

Is `tootctl accounts cull` another option to solve at least the size of accounts folder, here?

**Reply by ClearlyClaire** [reply]
> that's mostly on remote accounts for our instance has about 2500 accounts, only

Everything in `public/system/cache` is for remote content only.

> 162GB

162GB of media files sounds like a lot, but you do have a lot of remote media files either way.

> Is `tootctl accounts cull` another option to solve at least the size of accounts folder, here?

`tootctl accounts cull` might help a little, but it's unlikely to make a real difference; it's used to remove accounts that have been remotely deleted while your instance is unaware of that fact. In general, that should not happen, so I expect very few accounts to show up there.

`tootctl accounts prune` may be closer to what you want, but it's not in any released version yet.

**Reply by soletan** [reply]
Okay, thanks for the update. So I'll stick to the existing setup for now waiting for upcoming releases.

**Reply by soletan** [reply]
Minor update: after running tootctl media remove, the number of files last changed >7 days ago in media_attachments is 0 with consumption going from 162GB to 122GB, so everything seems to be working there and we are just grabbing a lot remote files all the time.

**Reply by melroy89** [reply]
Any other ideas after these years.. I'm now using 62.481 GB only for `/public/system/cache/accounts` (which is the largest folder in the cache directory).

**EDIT:** I think I will try: `tootctl media remove --prune-profiles --concurrency 15 --days 30` (this will actually clean up the biggest directory that is: `/system/cache/accounts`

**Comment by mygit67** [comment]
My backup plan excludes ../cache/*.
I just now had to restore a full machine backup because a) I couldn't logon to my single-user instance, and b) tootctl didn't work because the "database is in recovery mode", it said.
My server has 160GB disk space. But /cache(media) had grown to 139GB in 40 days  (since I had to cull the cache folder first to shrink my pay-as-you-go-backup size).  And today, server disk space was 95% full. 
So disk space was the reason. (And crontab hadn't done what I had set it up to.)

After the full restore (plus incremental) , I thought, I best bring the database up to speed regarding the now unavailable media. And ran: 

RAILS_ENV=production /home/mastodon/live/bin/tootctl preview_cards remove --days 1
RAILS_ENV=production /home/mastodon/live/bin/tootctl media remove --days 1
It reported to have deleted 2.34GB and 99GB respectively.  
I believe this to be a valid approach to heal the database after a cache-less machine restore. 

After logging into the instance as user, I can see most profile pictures from people I follow. 
Missing: Images in older toots in the columns of followed hashtags – and don't reload when I open the toot. 
Not missing: Images in older toots that I had boosted or faved. 

I can live with this. The restore took 50 minutes. Running the two tootcl jobs after a full restore took 15 minutes. YMMV.

**Reply by soletan** [reply]
Regarding crontab, I had issues with scripts failing to find ruby as it was locally installed in a user's folder. Thus, my crontab entries look more like this:

```
0 3 * * 1,4 cd /home/mastodon/live && PATH="/home/mastodon/.rbenv/shims:$PATH" RAILS_ENV=production bin/tootctl media remove
30 3 * * 1,4 cd /home/mastodon/live && PATH="/home/mastodon/.rbenv/shims:$PATH" RAILS_ENV=production bin/tootctl preview_cards remove
```

**Reply by mygit67** [reply]
Thanks. Is this your cronjob as user mastodon or as root?

**Reply by soletan** [reply]
It's the crontab of user mastodon.

**Reply by mygit67** [reply]
When you save crontab, does it also save it to a weird folder name, like tmp/crontab.osCliK/crontab  as for my user mastodon?

**Reply by soletan** [reply]
That's for backing up the editor session, only, making sure it won't accidentally corrupt the existing table. When it's valid on save, it should replace the existing crontab file of that user in /var/spool/cron/crontabs/mastodon. You shouldn't edit that file directly, that's why `crontab -e` has been made available.
